tests: bsim_bt: Extend connection test case.

wopu-ot opened the pull request:
Make peripheral role part of test case to build it with
split/non-split stack. Change connection parameters and channel map
after connection is established.

Signed-off-by: Wolfgang Puffitsch <wopu@oticon.com>

wopu-ot committed:
tests: bsim_bt: Make peripheral role part of test case.

Make peripheral role part of test case to build it with
split/non-split stack.

Signed-off-by: Wolfgang Puffitsch <wopu@oticon.com>

jhedberg commented on tests/bluetooth/bsim_bt/bsim_test_app/src/test_connect1.c:
```diff
@@ -268,8 +275,7 @@ static void device_found(const bt_addr_le_t *addr, s8_t rssi, u8_t type,
 	if (type == BT_LE_ADV_IND || type == BT_LE_ADV_DIRECT_IND) {
-		/* TODO: Move this to a place it can be shared */
-		ad_parse(ad, eir_found, (void *)addr);
+		bt_data_parse(ad, eir_found, (void *)addr);
```
This refactoring to use bt_data_parse() seems unrelated to the rest of the patch. Could you move that into a separate commit please? (can still be part of this same PR).

wopu-ot replied:
Yes, I will do that.

wopu-ot replied:
@jhedberg The commit has been split accordingly. Do you have any further comments?

aescolar commented on tests/bluetooth/bsim_bt/bsim_test_app/src/test_connect2.c:
```diff
@@ -0,0 +1,179 @@
+	/* Implement notification. At the moment there is no suitable way
+	 * of starting delayed work so we do it here
+	 */
+	for (int i = 0; i < WAIT_TIME-1; i++) {
```
It feels a bit too unreliable to have the test wait for WAIT_TIME-1 iterations before passing, as WAIT_TIME was just a timeout to fail the test.
Would it make sense to have this loop be an infinite loop, counting notifications sent,
and have the test pass after having sent a predefined number of notifications?

wopu-ot replied:
It makes sense, and I updated accordingly.

aescolar commented on tests/bluetooth/bsim_bt/bsim_test_app/tests_scripts/basic_conn.sh:
```diff
@@ -24,12 +24,13 @@ BOARD="${BOARD:-nrf52_bsim}"
-  -v=${VERBOSITY_LEVEL} -s=${SIMULATION_ID} -d=0 -RealEncryption=0
+Execute ./bs_${BOARD}_tests_bluetooth_bsim_bt_bsim_test_app_prj_conf \
+  -v=${VERBOSITY_LEVEL} -s=${SIMULATION_ID} -d=0 -RealEncryption=0 \
+  -testid=peripheral -rs=6
```
Seeding both devices with the same random seed seems a bit funny (same in the other testcase script below)

wopu-ot replied:
No need to do funny stuff, I use a different seed now for the peripheral.

aescolar commented on tests/bluetooth/bsim_bt/bsim_test_app/src/test_connect2.c:
```diff
@@ -0,0 +1,179 @@
+	printk("Testcase passed\n");
+
+	/* Wait to be killed */
+	for (;;) {
```
Doing it this way (in both sides) the devices will never end the simulation, but you rely on the sim_length passed to the Phy.
(This Phy timeout is 20 simulated seconds now)

If, in this device, all you want is to see it send 1 notification to pass, then you could have it change the test status to passed after the first notification, and just let this device go on sending notifications every second.
You could then let the other side device be the one that ends the simulation (with the same condition it had before).

aescolar commented on tests/bluetooth/bsim_bt/bsim_test_app/tests_scripts/basic_conn_ecrypted.sh:
```diff
@@ -24,12 +24,13 @@ BOARD="${BOARD:-nrf52_bsim}"
-Execute ./bs_${BOARD}_samples_bluetooth_peripheral_prj_conf \
-  -v=${VERBOSITY_LEVEL} -s=${SIMULATION_ID} -d=0 -RealEncryption=1
+Execute ./bs_${BOARD}_tests_bluetooth_bsim_bt_bsim_test_app_prj_conf \
+  -v=${VERBOSITY_LEVEL} -s=${SIMULATION_ID} -d=0 -RealEncryption=0 \
```
Ok.. it is not passing :)
```
d_01: @00:00:02.366132  [SUBSCRIBED]
d_00: @00:00:02.465774  [00:00:02.465,759] <err> bt_smp: reason 0x8
d_01: @00:00:03.041006  [NOTIFICATION] data 0x5674df4b length 2
d_01: @00:00:04.047255  [NOTIFICATION] data 0x5674df4b length 2
d_01: @00:00:04.047255  Testcase passed
d_00: @00:00:05.039520  Testcase passed
d_00: @00:00:09.122272  Disconnected (reason 8)
```

wopu-ot replied:
I reworked the logic to use callbacks, so encryption should work now.

wopu-ot committed:
tests: bsim_bt: Replace ad_parse function with bt_data_parse.

Use shared function, aligning with code in central_hr sample.

Signed-off-by: Wolfgang Puffitsch <wopu@oticon.com>

wopu-ot committed:
tests: bsim_bt: Extend connection test case.

Change connection parameters and channel map after connection is
established. Test encrypted connection also with split stack. Tighten
pass/fail criteria.

Signed-off-by: Wolfgang Puffitsch <wopu@oticon.com>

aescolar commented on tests/bluetooth/bsim_bt/bsim_test_app/src/test_connect2.c:
```diff
@@ -0,0 +1,162 @@
+
+	err = bt_enable(bt_ready);
+	if (err) {
+		printk("Bluetooth init failed (err %d)\n", err);
```
Should all these error conditions we treated as test case failures? (bst_result = Failed, and bs_trace_error_line() )

wopu-ot replied:
At least some of the error conditions made the test case fail implicitly by not proceeding. However, I now updated the pass/fail criteria to explicitly fail in case of error conditions and make this clear.

aescolar commented on tests/bluetooth/bsim_bt/bsim_test_app/src/test_connect1.c:
```diff
@@ -83,8 +97,18 @@ static u8_t notify_func(struct bt_conn *conn,
+
 		bst_result = Passed;
-		bs_trace_exit_time("Testcase passed\n");
+		printk("Testcase passed\n");
```
Silly question: Why this change?

aescolar commented on tests/bluetooth/bsim_bt/bsim_test_app/tests_scripts/basic_conn.sh:
```diff
@@ -24,12 +24,13 @@ BOARD="${BOARD:-nrf52_bsim}"
-  -v=${VERBOSITY_LEVEL} -s=${SIMULATION_ID} -d=0 -RealEncryption=0
+Execute ./bs_${BOARD}_tests_bluetooth_bsim_bt_bsim_test_app_prj_conf \
+  -v=${VERBOSITY_LEVEL} -s=${SIMULATION_ID} -d=0 -RealEncryption=0 \
+  -testid=peripheral -rs=23
```
Just a comment: You do not necessarily need to set the random seed to anything if you don't want. The random seed should be default be set to a value calculated from the device number (-d)

aescolar reviewed (changes requested):
It is way nicer now. The reject is only because I think failing the test on all those error checks would be worthwhile

jhedberg commented on tests/bluetooth/bsim_bt/bsim_test_app/src/test_connect2.c:
```diff
@@ -0,0 +1,172 @@
+#define WAIT_TIME 5 /*seconds*/
+extern enum bst_result_t bst_result;
+
+#define FAIL(...)					\
```
How come you're redefining these here? Wouldn't it make sense to have them in some local header file?

jhedberg reviewed (approved):
Looks ok to me, except the the seemingly unnecessary redefinition of PASS/FAIL (which I don't consider a blocker, so approving)
